contracts/layer-2/NFTBurner.sol

```diff
@@ -0,0 +1,163 @@
+import "@openzeppelin/contracts/interfaces/IERC721Receiver.sol";
+import "@openzeppelin/contracts/interfaces/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "solidity-bytes-utils/contracts/BytesLib.sol";
+
+/**
+ * @title NFTBurner
+ * @dev TODO
+ */
+contract NFTBurner is IERC721Receiver, AccessControl {
+    using SafeERC20 for IERC20;
+
+    address public constant DEAD_ADDRESS =
+        0x000000000000000000000000000000000000dEaD;
+
+    bytes32 public constant OPERATOR_ROLE = keccak256("OPERATOR_ROLE");
+
+    // Possible states for an existing token burn
+    bytes32 public constant BURN_RESERVED = keccak256("BURN_RESERVED");
+    bytes32 public constant BURN_EXECUTED = keccak256("BURN_EXECUTED");
+    bytes32 public constant BURN_REVERTED = keccak256("BURN_REVERTED");
+
+    struct Burn {
+        address from;
+        uint256[] tokenIds;
+        bytes32 state;
+    }
+
+    IERC721 public UltimateChampionsNFT;
+
+    // (keccak256 UID => Burn) mapping of burn operations
+    mapping(bytes32 => Burn) private _burns;
+
+    constructor(address nft) {
+        UltimateChampionsNFT = IERC721(nft);
+        _setupRole(DEFAULT_ADMIN_ROLE, _msgSender());
+    }
+
+    /**
+     * @dev See {IERC165-supportsInterface}.
+     */
+    function supportsInterface(bytes4 interfaceId)
+        public
+        view
+        virtual
+        override(AccessControl)
+        returns (bool)
+    {
+        return
+            interfaceId == type(IERC721Receiver).interfaceId ||
+            AccessControl.supportsInterface(interfaceId);
+    }
+
+    /**
+     * @dev See {IERC721Receiver-onERC721Received}.
+     */
+    function onERC721Received(
+        address operator,
+        address from,
+        uint256 tokenId,
+        bytes calldata data
+    ) external pure returns (bytes4) {
+        return this.onERC721Received.selector;
+    }
+
+    function getBurn(bytes32 UID) external view returns (Burn memory) {
+        return _burns[UID];
+    }
+
+    function isBurnReserved(bytes32 UID) public view returns (bool) {
+        return _burns[UID].state == BURN_RESERVED;
+    }
+
+    function isBurnProcessed(bytes32 UID) public view returns (bool) {
+        return
+            _burns[UID].state == BURN_EXECUTED ||
+            _burns[UID].state == BURN_REVERTED;
+    }
+
+    /**
+     * @dev sends a batch of NFCHAMP tokens from `from` to `to`.
+     * Requires this contract to be approved by the tokens' holder before hand.
+     */
+    function _batchTokenTransfer(
+        address from,
+        address to,
+        uint256[] memory tokenIds
+    ) private {
+        uint256 length = tokenIds.length;
+        for (uint256 i = 0; i < length; ) {
+            UltimateChampionsNFT.safeTransferFrom(from, to, tokenIds[i]);
+            unchecked {
+                ++i;
+            }
+        }
```

Comment:
We can limit the number of NFT to burn,this would allow to increment the variable in the loop

Reply:
This has to do with saving gas, which is more noticeable for larger loop runs. We don't need overflow checks for that variable since it can never overflow given that it is bounded by the maximum array length value (which will never overflow).